File: Core/Services/ReportExporter.swift
import Foundation

public enum ExportFormat: String, CaseIterable, Sendable, Hashable {
    case csv, pdf

    public var labelKey: String { "export.\(rawValue)" }
    public var fileExtension: String { rawValue }
    public var mimeType: String {
        switch self {
        case .csv: "text/csv"
        case .pdf: "application/pdf"
        }
    }
}

public protocol ReportExporter: Sendable {
    func exportAthletes(_ athletes: [Athlete], format: ExportFormat) -> Data
    func exportTournamentResults(
        tournament: Tournament,
        registrations: [TournamentRegistration],
        athletes: [Athlete],
        matches: [Match],
        format: ExportFormat
    ) -> Data
    func exportBranchPerformance(
        branches: [Branch],
        summaries: [BranchSummary],
        format: ExportFormat
    ) -> Data
    func exportMyData(
        user: User,
        athlete: Athlete?,
        registrations: [TournamentRegistration],
        matches: [Match]
    ) -> Data
}

/// CSV-only baseline implementation. PDF returns the same CSV bytes — the
/// platform-native PDF rendering lives in the SwiftUI layer where we can use
/// `ImageRenderer`. Keeping this pure-Foundation lets `Core/` stay UI-free.
public struct CSVReportExporter: ReportExporter {

    public init() {}

    public func exportAthletes(_ athletes: [Athlete], format: ExportFormat) -> Data {
        let header = "id,full_name,full_name_ar,branch_id,coach_id,belt,age,weight_kg,status\n"
        let rows = athletes.map { a in
            [
                a.id.uuidString,
                escape(a.fullName),
                escape(a.fullNameAr),
                a.branchID.uuidString,
                a.primaryCoachID?.uuidString ?? "",
                a.currentBelt.label,
                String(a.age),
                String(a.weightKg),
                a.status.rawValue
            ].joined(separator: ",")
        }.joined(separator: "\n")
        return (header + rows).data(using: .utf8) ?? Data()
    }

    public func exportTournamentResults(
        tournament: Tournament,
        registrations: [TournamentRegistration],
        athletes: [Athlete],
        matches: [Match],
        format: ExportFormat
    ) -> Data {
        let athleteByID = Dictionary(uniqueKeysWithValues: athletes.map { ($0.id, $0) })
        var lines: [String] = []
        lines.append("# tournament: \(escape(tournament.name))")
        lines.append("# starts_at: \(ISO8601DateFormatter().string(from: tournament.startsAt))")
        lines.append("")
        lines.append("athlete,weight_category,seed,status")
        for r in registrations {
            let name = athleteByID[r.athleteID]?.fullName ?? r.athleteID.uuidString
            lines.append([escape(name), r.weightCategory.rawValue, r.seedRank.map { String($0) } ?? "", r.status.rawValue].joined(separator: ","))
        }
        lines.append("")
        lines.append("athlete,opponent,our_score,opp_score,won,medal,date")
        let df = ISO8601DateFormatter()
        for m in matches {
            let athleteName = athleteByID[m.ourAthleteID]?.fullName ?? m.ourAthleteID.uuidString
            lines.append([
                escape(athleteName),
                escape(m.opponentName ?? "—"),
                String(m.ourScore),
                String(m.opponentScore),
                m.won ? "1" : "0",
                m.medal.rawValue,
                df.string(from: m.date)
            ].joined(separator: ","))
        }
        return lines.joined(separator: "\n").data(using: .utf8) ?? Data()
    }

    public func exportBranchPerformance(
        branches: [Branch],
        summaries: [BranchSummary],
        format: ExportFormat
    ) -> Data {
        let header = "code,name,name_ar,athletes,utilisation,composite,grade\n"
        let rows = summaries.map { s in
            [
                s.branch.code,
                escape(s.branch.name),
                escape(s.branch.nameAr),
                String(s.athleteCount),
                String(format: "%.2f", s.utilisation),
                String(format: "%.1f", s.composite),
                s.grade.label
            ].joined(separator: ",")
        }.joined(separator: "\n")
        return (header + rows).data(using: .utf8) ?? Data()
    }

    public func exportMyData(
        user: User,
        athlete: Athlete?,
        registrations: [TournamentRegistration],
        matches: [Match]
    ) -> Data {
        struct Bundle: Codable {
            let user: User
            let athlete: Athlete?
            let registrations: [TournamentRegistration]
            let matches: [Match]
            let exportedAt: Date
        }
        let bundle = Bundle(
            user: user,
            athlete: athlete,
            registrations: registrations,
            matches: matches,
            exportedAt: Date()
        )
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return (try? encoder.encode(bundle)) ?? Data()
    }

    private func escape(_ s: String) -> String {
        if s.contains(",") || s.contains("\"") || s.contains("\n") {
            return "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\""
        }
        return s
    }
}

File: flutter_app/lib/core/services/report_exporter.dart
import 'dart:convert';

import '../models/athlete.dart';
import '../models/branch.dart';
import '../models/entity_id.dart';
import '../models/match.dart';
import '../models/tournament.dart';
import '../models/user.dart';
import 'score_engine.dart';

/// 1:1 port of Core/Services/ReportExporter.swift.
///
/// Pure CSV / JSON export logic. PDF generation requires platform rendering
/// (`ImageRenderer` in SwiftUI, a PDF package in Flutter) — see the
/// TODO(platform) note on [CsvReportExporter.exportBranchPerformance].
///
/// All methods return a [String] (the CSV / JSON payload). The platform layer
/// is responsible for writing to a file, sharing via the share-sheet, etc.

enum ExportFormat {
  csv,
  pdf;

  String get labelKey => 'export.$name';
  String get fileExtension => name;
  String get mimeType => switch (this) {
    ExportFormat.csv => 'text/csv',
    ExportFormat.pdf => 'application/pdf',
  };
}

/// Lightweight summary of one branch used by the report exporter. Mirrors
/// `BranchSummary` from Core/Stores/BranchesStore.swift (which lives in the
/// store layer of the Swift app).
class BranchSummary {
  final EntityID id;
  final Branch branch;
  final double composite;
  final LetterGrade grade;
  final int athleteCount;
  final double utilisation;

  const BranchSummary({
    required this.id,
    required this.branch,
    required this.composite,
    required this.grade,
    required this.athleteCount,
    required this.utilisation,
  });
}

/// Contract for all report exporters. Returns a UTF-8 [String] (CSV or JSON).
/// TODO(platform): the platform layer converts the String to bytes, writes
/// to a temp file, and triggers the system share-sheet.
abstract class ReportExporter {
  String exportAthletes(List<Athlete> athletes, ExportFormat format);

  String exportTournamentResults({
    required Tournament tournament,
    required List<TournamentRegistration> registrations,
    required List<Athlete> athletes,
    required List<Match> matches,
    ExportFormat format,
  });

  String exportBranchPerformance({
    required List<Branch> branches,
    required List<BranchSummary> summaries,
    ExportFormat format,
  });

  String exportMyData({
    required User user,
    Athlete? athlete,
    required List<TournamentRegistration> registrations,
    required List<Match> matches,
  });
}

/// CSV-only baseline implementation. PDF returns the same CSV payload —
/// the platform-native PDF rendering lives in the widget layer.
/// Keeping this pure-Dart ensures lib/core stays UI-free.
class CsvReportExporter implements ReportExporter {
  const CsvReportExporter();

  @override
  String exportAthletes(List<Athlete> athletes, ExportFormat format) {
    const header =
        'id,full_name,full_name_ar,branch_id,coach_id,belt,age,weight_kg,status\n';
    final rows = athletes
        .map((a) {
          return [
            a.id,
            _escape(a.fullName),
            _escape(a.fullNameAr),
            a.branchId,
            a.primaryCoachId ?? '',
            a.currentBelt.label,
            a.age.toString(),
            a.weightKg.toString(),
            a.status.name,
          ].join(',');
        })
        .join('\n');
    return header + rows;
  }

  @override
  String exportTournamentResults({
    required Tournament tournament,
    required List<TournamentRegistration> registrations,
    required List<Athlete> athletes,
    required List<Match> matches,
    ExportFormat format = ExportFormat.csv,
  }) {
    final athleteById = Map.fromEntries(athletes.map((a) => MapEntry(a.id, a)));
    final lines = <String>[];
    lines.add('# tournament: ${_escape(tournament.name)}');
    lines.add('# starts_at: ${tournament.startsAt.toIso8601String()}');
    lines.add('');
    lines.add('athlete,weight_category,seed,status');
    for (final r in registrations) {
      final name = athleteById[r.athleteId]?.fullName ?? r.athleteId;
      lines.add(
        [
          _escape(name),
          r.weightCategory.name,
          r.seedRank?.toString() ?? '',
          r.status.name,
        ].join(','),
      );
    }
    lines.add('');
    lines.add('athlete,opponent,our_score,opp_score,won,medal,date');
    for (final m in matches) {
      final athleteName =
          athleteById[m.ourAthleteId]?.fullName ?? m.ourAthleteId;
      lines.add(
        [
          _escape(athleteName),
          _escape(m.opponentName ?? '—'),
          m.ourScore.toString(),
          m.opponentScore.toString(),
          m.won ? '1' : '0',
          m.medal.name,
          m.date.toIso8601String(),
        ].join(','),
      );
    }
    return lines.join('\n');
  }

  @override
  String exportBranchPerformance({
    required List<Branch> branches,
    required List<BranchSummary> summaries,
    ExportFormat format = ExportFormat.csv,
  }) {
    // TODO(platform): PDF rendering is deferred — implement in the platform
    // stage using a PDF generation package (e.g. pdf/printing).
    const header = 'code,name,name_ar,athletes,utilisation,composite,grade\n';
    final rows = summaries
        .map((s) {
          return [
            s.branch.code,
            _escape(s.branch.name),
            _escape(s.branch.nameAr),
            s.athleteCount.toString(),
            s.utilisation.toStringAsFixed(2),
            s.composite.toStringAsFixed(1),
            s.grade.label,
          ].join(',');
        })
        .join('\n');
    return header + rows;
  }

  @override
  String exportMyData({
    required User user,
    Athlete? athlete,
    required List<TournamentRegistration> registrations,
    required List<Match> matches,
  }) {
    final bundle = {
      'user': user.toJson(),
      'athlete': athlete?.toJson(),
      'registrations': registrations.map((r) => r.toJson()).toList(),
      'matches': matches.map((m) => m.toJson()).toList(),
      'exportedAt': DateTime.now().toIso8601String(),
    };
    const encoder = JsonEncoder.withIndent('  ');
    return encoder.convert(bundle);
  }

  String _escape(String s) {
    if (s.contains(',') || s.contains('"') || s.contains('\n')) {
      return '"${s.replaceAll('"', '""')}"';
    }
    return s;
  }
}
